src/Cimpress.Auth0.Client/Auth0v2TokenProvider.cs

```diff
@@ -0,0 +1,124 @@
+        /// </summary>
+        /// <param name="defaultSettings">The settings.</param>
+        /// <param name="autoScheduler">The auto-scheduler that refreshes the Auth0 token after X minutes.</param>
+        public Auth0v2TokenProvider(ILoggerFactory loggerFactory, Auth0ClientSettings defaultSettings, IAuthenticationApiClient authenticationApiClient = null, IAutoScheduler autoScheduler = null) : base(loggerFactory, defaultSettings, authenticationApiClient)
+        {
+            this.autoScheduler = autoScheduler ?? new AutoScheduler(loggerFactory, this);
+            defaultClientSecret = defaultSettings.Auth0ClientSecret;
+            defaultAudience = defaultSettings.Auth0Audience;
+        }
+
+        public override void ScheduleAutoRefresh(Auth0ClientSettings auth0ClientSettings)
+        {
+            autoScheduler.ScheduleRefresh(auth0ClientSettings);
+        }
+
+        /// <summary>
+        /// Locally caches Auth0 settings for a given client based on the <paramref name="settings"/> provided.
+        /// If the <paramref name="settings"/> instance doesn't provide certain parameters, the provider falls
+        /// back to the submitted default settings.
+        /// </summary>
+        public override void CacheAuthSettings(Auth0ClientSettings settings)
+        {
+            // apply defaults
+            settings.Auth0ClientSecret = string.IsNullOrWhiteSpace(settings.Auth0ClientSecret)
+                ? defaultClientSecret
+                : settings.Auth0ClientSecret;
+            settings.Auth0Audience = string.IsNullOrWhiteSpace(settings.Auth0Audience)
+                ? defaultAudience
+                : settings.Auth0Audience;
+            settings.Auth0ServerUrl = string.IsNullOrWhiteSpace(settings.Auth0ServerUrl)
+                ? defaultDomain
+                : settings.Auth0ServerUrl;
+            settings.Auth0RefreshToken = string.IsNullOrWhiteSpace(settings.Auth0RefreshToken)
+                ? defaultRefreshToken
+                : settings.Auth0RefreshToken;
+            settings.AutoRefreshAfter = settings.AutoRefreshAfter == TimeSpan.MinValue
+                ? defaultAutoRefreshAfter
+                : settings.AutoRefreshAfter;
+
+            // cache settings
+            clientTokenCache.TryAdd(settings.Auth0ClientId, settings);
+        }
+
+        /// <summary>
+        /// Updates the auth0 authentication header for the client id using username and password asynchronous.
+        /// </summary>
+        /// <param name="clientId">The client identifier.</param>
+        /// <param name="forceRefresh">if set to <c>true</c> [force refresh].</param>
+        /// <returns>
+        /// A task, when completed, ensures that the authentication header got updated.
+        /// </returns>
+        /// <exception cref="System.Collections.Generic.KeyNotFoundException"></exception>
+        public override async Task UpdateAuthHeaderWithCredentialsAsync(string clientId, bool forceRefresh = false)
```

Comment:
There are a couple of methods are all fundamentally the same since they are wrappers, but call only one function differently.  Did you have any thoughts on how to reuse the wrapper code (which caches the resultant token)?